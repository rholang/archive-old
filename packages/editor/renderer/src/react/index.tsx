import * as React from 'react';
// @ts-ignore: unused variable
// prettier-ignore
import { ComponentType, Consumer, Provider } from 'react';
import { Fragment, Mark, MarkType, Node, Schema } from 'prosemirror-model';
import { Serializer } from '../';
import { RendererAppearance } from '../ui/Renderer/types';
import { AnalyticsEventPayload } from '../analytics/events';

import {
  Doc,
  mergeTextNodes,
  isTextWrapper,
  TextWrapper,
  toReact,
} from './nodes';

import { toReact as markToReact } from './marks';
import {
  ProviderFactory,
  getMarksByOrder,
  isSameMark,
  EventHandlers,
  ExtensionHandlers,
  calcTableColumnWidths,
} from '@atlaskit/editor-common';
import { getText } from '../utils';

export interface RendererContext {
  objectAri?: string;
  containerAri?: string;
  adDoc?: any;
  schema?: Schema;
}

export interface ConstructorParams {
  providers?: ProviderFactory;
  eventHandlers?: EventHandlers;
  extensionHandlers?: ExtensionHandlers;
  portal?: HTMLElement;
  objectContext?: RendererContext;
  appearance?: RendererAppearance;
  disableHeadingIDs?: boolean;
  allowDynamicTextSizing?: boolean;
  allowHeadingAnchorLinks?: boolean;
  allowColumnSorting?: boolean;
  fireAnalyticsEvent?: (event: AnalyticsEventPayload) => void;
  shouldOpenMediaViewer?: boolean;
}

type MarkWithContent = Partial<Mark<any>> & {
  content: Array<MarkWithContent | Node<any>>;
};

function mergeMarks(marksAndNodes: Array<MarkWithContent | Node>) {
  return marksAndNodes.reduce(
    (acc, markOrNode) => {
      const prev = (acc.length && acc[acc.length - 1]) || null;

      if (
        markOrNode.type instanceof MarkType &&
        prev &&
        prev.type instanceof MarkType &&
        Array.isArray(prev.content) &&
        isSameMark(prev as Mark, markOrNode as Mark)
      ) {
        prev.content = mergeMarks(
          prev.content.concat((markOrNode as MarkWithContent).content),
        );
      } else {
        acc.push(markOrNode);
      }

      return acc;
    },
    [] as Array<MarkWithContent | Node>,
  );
}

export default class ReactSerializer implements Serializer<JSX.Element> {
  private providers?: ProviderFactory;
  private eventHandlers?: EventHandlers;
  private extensionHandlers?: ExtensionHandlers;
  private portal?: HTMLElement;
  private rendererContext?: RendererContext;
  private appearance?: RendererAppearance;
  private disableHeadingIDs?: boolean;
  private headingIds: string[] = [];
  private allowDynamicTextSizing?: boolean;
  private allowHeadingAnchorLinks?: boolean;
  private allowColumnSorting?: boolean;
  private fireAnalyticsEvent?: (event: AnalyticsEventPayload) => void;
  private shouldOpenMediaViewer?: boolean;

  constructor({
    providers,
    eventHandlers,
    extensionHandlers,
    portal,
    objectContext,
    appearance,
    disableHeadingIDs,
    allowDynamicTextSizing,
    allowHeadingAnchorLinks,
    allowColumnSorting,
    fireAnalyticsEvent,
    shouldOpenMediaViewer,
  }: ConstructorParams) {
    this.providers = providers;
    this.eventHandlers = eventHandlers;
    this.extensionHandlers = extensionHandlers;
    this.portal = portal;
    this.rendererContext = objectContext;
    this.appearance = appearance;
    this.disableHeadingIDs = disableHeadingIDs;
    this.allowDynamicTextSizing = allowDynamicTextSizing;
    this.allowHeadingAnchorLinks = allowHeadingAnchorLinks;
    this.allowColumnSorting = allowColumnSorting;
    this.fireAnalyticsEvent = fireAnalyticsEvent;
    this.shouldOpenMediaViewer = shouldOpenMediaViewer;
  }

  private resetState() {
    this.headingIds = [];
  }

  serializeFragment(
    fragment: Fragment,
    props: any = {},
    target: any = Doc,
    key: string = 'root-0',
    parentInfo?: { parentIsIncompleteTask: boolean; path: Array<Node> },
  ): JSX.Element | null {
    // This makes sure that we reset internal state on re-render.
    if (key === 'root-0') {
      this.resetState();
    }

    const content = ReactSerializer.getChildNodes(fragment).map(
      (node, index) => {
        if (isTextWrapper(node)) {
          return this.serializeTextWrapper(node.content);
        }

        let props;

        if (node.type.name === 'table') {
          props = this.getTableProps(node);
        } else if (node.type.name === 'date') {
          props = this.getDateProps(node, parentInfo);
        } else if (node.type.name === 'heading') {
          props = this.getHeadingProps(node, parentInfo && parentInfo.path);
        } else if (['tableHeader', 'tableRow'].indexOf(node.type.name) > -1) {
          props = this.getTableChildrenProps(node);
        } else if (node.type.name === 'media') {
          props = this.getMediaProps(node);
        } else {
          props = this.getProps(node);
        }

        let currentPath = (parentInfo && parentInfo.path) || [];

        const parentIsIncompleteTask =
          node.type.name === 'taskItem' && node.attrs.state !== 'DONE';

        let pInfo = {
          parentIsIncompleteTask,
          path: [...currentPath, node],
        };

        const serializedContent = this.serializeFragment(
          node.content,
          props,
          toReact(node),
          `${node.type.name}-${index}`,
          pInfo,
        );

        if (node.marks && node.marks.length) {
          return ([] as Array<Mark>)
            .concat(node.marks)
            .reverse()
            .reduce((acc, mark) => {
              return this.renderMark(
                markToReact(mark),
                this.getMarkProps(mark),
                `${mark.type.name}-${index}`,
                acc,
              );
            }, serializedContent);
        }

        return serializedContent;
      },
    );

    return this.renderNode(target, props, key, content);
  }

  private serializeTextWrapper(content: Node[]) {
    return ReactSerializer.buildMarkStructure(content).map((mark, index) =>
      this.serializeMark(mark, index),
    );
  }

  private serializeMark(mark: Mark, index: number = 0) {
    if (mark.type.name === 'text') {
      return (mark as any).text;
    }

    const content = ((mark as any).content || []).map(
      (child: Mark, index: number) => this.serializeMark(child, index),
    );
    return this.renderMark(
      markToReact(mark),
      this.getMarkProps(mark),
      `${mark.type.name}-${index}`,
      content,
    );
  }

  private renderNode(
    NodeComponent: ComponentType<any>,
    props: any,
    key: string,
    content: string | JSX.Element | any[] | null | undefined,
  ): JSX.Element {
    return (
      <NodeComponent key={key} {...props}>
        {content}
      </NodeComponent>
    );
  }

  private renderMark(
    MarkComponent: ComponentType<any>,
    props: any,
    key: string,
    content: any,
  ) {
    return (
      <MarkComponent key={key} {...props}>
        {content}
      </MarkComponent>
    );
  }

  private getTableChildrenProps(node: Node) {
    return {
      ...this.getProps(node),
      allowColumnSorting: this.allowColumnSorting,
    };
  }

  private getTableProps(node: Node) {
    return {
      ...this.getProps(node),
      allowColumnSorting: this.allowColumnSorting,
      columnWidths: calcTableColumnWidths(node),
      tableNode: node,
    };
  }

  private getDateProps(
    node: Node,
    parentInfo: { parentIsIncompleteTask: boolean } | undefined,
  ) {
    return {
      timestamp: node.attrs && node.attrs.timestamp,
      parentIsIncompleteTask: parentInfo && parentInfo.parentIsIncompleteTask,
    };
  }
  private getMediaProps(node: Node) {
    return {
      ...this.getProps(node),
      shouldOpenMediaViewer: this.shouldOpenMediaViewer,
    };
  }

  private getProps(node: Node) {
    return {
      text: node.text,
      providers: this.providers,
      eventHandlers: this.eventHandlers,
      extensionHandlers: this.extensionHandlers,
      portal: this.portal,
      rendererContext: this.rendererContext,
      serializer: this,
      content: node.content ? node.content.toJSON() : undefined,
      allowDynamicTextSizing: this.allowDynamicTextSizing,
      allowHeadingAnchorLinks: this.allowHeadingAnchorLinks,
      rendererAppearance: this.appearance,
      fireAnalyticsEvent: this.fireAnalyticsEvent,
      ...node.attrs,
    };
  }

  private headingAnchorSupported(path: Array<Node> = []): boolean {
    return (
      path.length === 0 || path[path.length - 1].type.name === 'layoutColumn'
    );
  }

  private getHeadingProps(node: Node, path: Array<Node> = []) {
    return {
      ...node.attrs,
      content: node.content ? node.content.toJSON() : undefined,
      headingId: this.getHeadingId(node),
      showAnchorLink:
        this.allowHeadingAnchorLinks &&
        !this.disableHeadingIDs &&
        this.headingAnchorSupported(path),
    };
  }

  // The return value of this function is NOT url encoded,
  // In HTML5 standard, id can contain any characters, encoding is no necessary.
  // Plus we trying to avoid double encoding, therefore we leave the value as is.
  // Remember to use encodeURIComponent when generating url from the id value.
  private getHeadingId(node: Node) {
    if (this.disableHeadingIDs || !node.content.size) {
      return;
    }

    // We are not use node.textContent here, because we would like to handle cases where
    // headings only contain inline blocks like emoji, status and date.
    const nodeContent = (node as any).content
      .toJSON()
      .reduce((acc: string, node: any) => acc.concat(getText(node) || ''), '')
      .trim()
      .replace(/\s/g, '-');

    if (!nodeContent) {
      return;
    }

    return this.getUniqueHeadingId(nodeContent);
  }

  private getUniqueHeadingId(baseId: string, counter = 0): string {
    if (counter === 0 && this.headingIds.indexOf(baseId) === -1) {
      this.headingIds.push(baseId);
      return baseId;
    } else if (counter !== 0) {
      const headingId = `${baseId}.${counter}`;
      if (this.headingIds.indexOf(headingId) === -1) {
        this.headingIds.push(headingId);
        return headingId;
      }
    }

    return this.getUniqueHeadingId(baseId, ++counter);
  }

  private getMarkProps(mark: Mark): any {
    const { key, ...otherAttrs } = mark.attrs;
    return {
      eventHandlers: this.eventHandlers,
      markKey: key,
      ...otherAttrs,
    };
  }

  static getChildNodes(fragment: Fragment): (Node | TextWrapper)[] {
    const children: Node[] = [];
    fragment.forEach(node => {
      children.push(node);
    });
    return mergeTextNodes(children) as Node[];
  }

  static getMarks(node: Node): Mark[] {
    if (!node.marks || node.marks.length === 0) {
      return [];
    }

    return getMarksByOrder(node.marks);
  }

  static buildMarkStructure(content: Node[]) {
    return mergeMarks(
      content.map(node => {
        const nodeMarks = this.getMarks(node);
        if (nodeMarks.length === 0) {
          return node;
        }

        return nodeMarks.reverse().reduce(
          (acc, mark) => {
            const { eq } = mark;

            return {
              ...mark,
              eq,
              content: [acc],
            };
          },
          node as any,
        );
      }),
    ) as Mark[];
  }

  static fromSchema(
    _schema: Schema,
    {
      providers,
      eventHandlers,
      extensionHandlers,
      appearance,
      disableHeadingIDs,
      allowDynamicTextSizing,
      allowHeadingAnchorLinks,
      allowColumnSorting,
    }: ConstructorParams,
  ): ReactSerializer {
    // TODO: Do we actually need the schema here?
    return new ReactSerializer({
      providers,
      eventHandlers,
      extensionHandlers,
      appearance,
      disableHeadingIDs,
      allowDynamicTextSizing,
      allowHeadingAnchorLinks,
      allowColumnSorting,
    });
  }
}
