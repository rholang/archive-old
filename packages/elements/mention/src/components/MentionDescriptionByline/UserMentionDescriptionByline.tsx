import * as React from 'react';
import { DescriptionBylineProps } from './types';
import { renderHighlight } from '../MentionItem/MentionHighlightHelpers';
import { DescriptionBylineStyle } from './styles';

export default class UserMentionDescriptionByline extends React.PureComponent<
  DescriptionBylineProps,
  {}
> {
  render() {
    const { highlight, name, nickname } = this.props.mention;
    const nicknameHighlights = highlight && highlight.nickname;

    if (name === nickname) {
      return null;
    }

    return renderHighlight(
      DescriptionBylineStyle,
      nickname,
      nicknameHighlights,
      '@',
    );
  }
}
