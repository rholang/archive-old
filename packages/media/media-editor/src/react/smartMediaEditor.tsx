import * as React from 'react';
import uuidV4 from 'uuid/v4';
import { Subscription } from 'rxjs/Subscription';

import {
  intlShape,
  IntlProvider,
  injectIntl,
  InjectedIntlProps,
} from 'react-intl';

import {
  withAnalyticsContext,
  withAnalyticsEvents,
  WithAnalyticsEventsProps,
} from '@atlaskit/analytics-next';

import {
  MediaClient,
  UploadableFile,
  FileIdentifier,
} from '@atlaskit/media-client';
// Importing from own entry-point, since we dont' want to bring whole media-client at this point
import { RECENTS_COLLECTION } from '@atlaskit/media-client/constants';
import { messages, Shortcut } from '@atlaskit/media-ui';
import ModalDialog, { ModalTransition } from '@atlaskit/modal-dialog';
import Spinner from '@atlaskit/spinner';

import EditorView from './editorView/editorView';
import { Blanket, SpinnerWrapper } from './styled';
import { fileToBase64, fireAnalyticsEvent } from '../util';
import ErrorView from './editorView/errorView/errorView';
import { CancelInputType, Dimensions, ShapeParameters, Tool } from '../common';

import {
  name as packageName,
  version as packageVersion,
} from '../version.json';

export const convertFileNameToPng = (fileName?: string) => {
  if (!fileName) {
    return 'annotated-image.png';
  }
  if (fileName.endsWith('.png')) {
    return fileName;
  } else {
    if (fileName.lastIndexOf('.') === 0 || fileName.lastIndexOf('.') === -1) {
      return `${fileName}.png`;
    } else {
      return `${fileName.substring(0, fileName.lastIndexOf('.'))}.png`;
    }
  }
};

export interface SmartMediaEditorProps {
  identifier: FileIdentifier;
  mediaClient: MediaClient;
  onUploadStart?: (identifier: FileIdentifier, dimensions: Dimensions) => void;
  onFinish?: (identifier: FileIdentifier) => void;
  onClose?: () => void;
}

export interface SmartMediaEditorState {
  hasError: boolean;
  errorMessage?: any;
  imageUrl?: string;
  hasBeenEdited: boolean;
  closeIntent: boolean;
}

export class SmartMediaEditor extends React.Component<
  SmartMediaEditorProps & InjectedIntlProps & WithAnalyticsEventsProps,
  SmartMediaEditorState
> {
  fileName?: string;
  state: SmartMediaEditorState = {
    hasError: false,
    hasBeenEdited: false,
    closeIntent: false,
  };
  getFileSubscription?: Subscription;
  uploadFileSubscription?: Subscription;

  static contextTypes = {
    intl: intlShape,
  };

  private getFileUnsubscribeTimeoutId: number | undefined;
  private uploadFileUnsubscribeTimeoutId: number | undefined;

  componentDidMount() {
    const { identifier } = this.props;
    this.getFile(identifier);
  }

  UNSAFE_componentWillReceiveProps(nextProps: Readonly<SmartMediaEditorProps>) {
    const { identifier, mediaClient } = this.props;
    if (
      nextProps.identifier.id !== identifier.id ||
      nextProps.mediaClient !== mediaClient
    ) {
      this.getFile(nextProps.identifier);
    }
  }

  componentWillUnmount() {
    window.clearTimeout(this.getFileUnsubscribeTimeoutId);
    window.clearTimeout(this.uploadFileUnsubscribeTimeoutId);

    const { getFileSubscription, uploadFileSubscription } = this;
    if (getFileSubscription) {
      getFileSubscription.unsubscribe();
    }
    if (uploadFileSubscription) {
      uploadFileSubscription.unsubscribe();
    }
  }

  getFile = async (identifier: FileIdentifier) => {
    const { mediaClient } = this.props;
    const { collectionName, occurrenceKey } = identifier;
    const id = await identifier.id;
    const getFileSubscription = mediaClient.file
      .getFileState(id, { collectionName, occurrenceKey })
      .subscribe({
        next: async state => {
          if (state.status === 'error') {
            this.onError(state.message);
            this.getFileUnsubscribeTimeoutId = window.setTimeout(
              () => getFileSubscription.unsubscribe(),
              0,
            );
            return;
          }

          const { name, preview, status } = state;
          this.fileName = name;

          if (status === 'processed') {
            this.setRemoteImageUrl(identifier);
            this.getFileUnsubscribeTimeoutId = window.setTimeout(
              () => getFileSubscription.unsubscribe(),
              0,
            );
          } else if (preview) {
            const { value } = await preview;
            if (value instanceof Blob) {
              const imageUrl = await fileToBase64(value);
              this.setState({
                imageUrl,
              });
            } else {
              this.setState({
                imageUrl: value,
              });
            }

            this.getFileUnsubscribeTimeoutId = window.setTimeout(
              () => getFileSubscription.unsubscribe(),
              0,
            );
          }
        },
        error: error => {
          this.onError(error);
        },
      });
    this.getFileSubscription = getFileSubscription;
  };

  setRemoteImageUrl = async (identifier: FileIdentifier) => {
    const { mediaClient } = this.props;
    const id = await identifier.id;
    const imageUrl = await mediaClient.getImageUrl(id, {
      collection: identifier.collectionName,
      mode: 'full-fit',
    });
    this.setState({
      imageUrl,
    });
  };

  copyFileToUserCollection = async (fileId: string) => {
    const {
      mediaClient: {
        config: { userAuthProvider, authProvider },
        file,
      },
      identifier: { collectionName },
    } = this.props;

    if (userAuthProvider) {
      const source = {
        id: fileId,
        collection: collectionName,
        authProvider,
      };
      const destination = {
        collection: RECENTS_COLLECTION,
        authProvider: userAuthProvider,
        occurrenceKey: uuidV4(),
      };
      await file.copyFile(source, destination);
    }
  };

  private onSave = (imageData: string, dimensions: Dimensions) => {
    const { fileName } = this;
    const {
      mediaClient,
      identifier,
      onUploadStart,
      onFinish,
      intl: { formatMessage },
    } = this.props;
    const { hasBeenEdited } = this.state;

    const { collectionName } = identifier;
    const uploadableFile: UploadableFile = {
      content: imageData,
      collection: collectionName,
      name: convertFileNameToPng(fileName),
    };
    const id = uuidV4();
    const occurrenceKey = uuidV4();
    const touchedFiles = mediaClient.file.touchFiles(
      [
        {
          fileId: id,
          collection: collectionName,
          occurrenceKey,
        },
      ],
      collectionName,
    );
    const deferredUploadId = touchedFiles.then(
      touchedFiles => touchedFiles.created[0].uploadId,
    );
    const uploadableFileUpfrontIds = {
      id,
      deferredUploadId,
      occurrenceKey,
    };

    const uploadingFileState = mediaClient.file.upload(
      uploadableFile,
      undefined,
      uploadableFileUpfrontIds,
    );
    const newFileIdentifier: FileIdentifier = {
      id,
      collectionName,
      mediaItemType: 'file',
      occurrenceKey,
    };
    const timeStarted = Date.now();

    fireAnalyticsEvent(
      {
        eventType: 'ui',
        action: 'clicked',
        actionSubject: 'button',
        actionSubjectId: 'saveButton',
        attributes: {
          annotated: hasBeenEdited,
        },
      },
      this.props.createAnalyticsEvent,
    );

    const uploadingFileStateSubscription = uploadingFileState.subscribe({
      next: fileState => {
        if (fileState.status === 'processing') {
          this.copyFileToUserCollection(fileState.id).then(() => {
            if (onFinish) {
              onFinish(newFileIdentifier);
            }

            fireAnalyticsEvent(
              {
                eventType: 'track',
                action: 'uploaded',
                actionSubject: 'media',
                actionSubjectId: id,
                attributes: {
                  status: 'success',
                  fileStatus: fileState.status,
                  fileMediatype: fileState.mediaType,
                  fileMimetype: fileState.mimeType,
                  fileSize: fileState.size,
                  uploadDurationMsec: Date.now() - timeStarted,
                  annotated: hasBeenEdited,
                },
              },
              this.props.createAnalyticsEvent,
            );

            this.uploadFileUnsubscribeTimeoutId = window.setTimeout(
              () => uploadingFileStateSubscription.unsubscribe(),
              0,
            );
          });
        } else if (
          fileState.status === 'failed-processing' ||
          fileState.status === 'error'
        ) {
          this.onError(formatMessage(messages.could_not_save_image));
          this.uploadFileUnsubscribeTimeoutId = window.setTimeout(
            () => uploadingFileStateSubscription.unsubscribe(),
            0,
          );

          fireAnalyticsEvent(
            {
              eventType: 'track',
              action: 'uploaded',
              actionSubject: 'media',
              actionSubjectId: id,
              attributes: {
                status: 'fail',
                failReason: formatMessage(messages.could_not_save_image),
                fileStatus: fileState.status,
                uploadDurationMsec: Date.now() - timeStarted,
                annotated: hasBeenEdited,
              },
            },
            this.props.createAnalyticsEvent,
          );
        }
      },
    });
    if (onUploadStart) {
      onUploadStart(newFileIdentifier, dimensions);
    }
  };

  private onAnyEdit = (tool: Tool, shapeParameters: ShapeParameters) => {
    const { hasBeenEdited } = this.state;

    fireAnalyticsEvent(
      {
        eventType: 'ui',
        action: 'annotated',
        actionSubject: 'annotation',
        actionSubjectId: tool,
        attributes: shapeParameters,
      },
      this.props.createAnalyticsEvent,
    );

    if (!hasBeenEdited) {
      this.setState({ hasBeenEdited: true });
    }
  };

  private closeConfirmationDialog = () => {
    this.setState({ closeIntent: false });
  };

  private closeAnyway = () => {
    const { onClose } = this.props;
    const { hasBeenEdited } = this.state;
    this.closeConfirmationDialog();

    fireAnalyticsEvent(
      {
        eventType: 'ui',
        action: 'clicked',
        actionSubject: 'button',
        actionSubjectId: 'confirmCancelButton',
        attributes: {
          annotated: hasBeenEdited,
        },
      },
      this.props.createAnalyticsEvent,
    );

    if (onClose) {
      onClose();
    }
  };

  private renderDeleteConfirmation = () => {
    const {
      intl: { formatMessage },
    } = this.props;
    const { closeIntent } = this.state;

    if (closeIntent) {
      const actions = [
        {
          text: formatMessage(messages.annotate_confirmation_close_anyway),
          onClick: this.closeAnyway,
        },
        {
          text: formatMessage(messages.cancel),
          onClick: this.closeConfirmationDialog,
        },
      ];
      return (
        <ModalTransition>
          <ModalDialog
            width="small"
            appearance="danger"
            heading={formatMessage(messages.annotate_confirmation_heading)}
            actions={actions}
            onClose={this.closeConfirmationDialog}
          >
            {formatMessage(messages.annotate_confirmation_content)}
          </ModalDialog>
        </ModalTransition>
      );
    }
    return null;
  };

  onCancel = (input: CancelInputType) => {
    const { hasBeenEdited } = this.state;
    const { onClose } = this.props;

    fireAnalyticsEvent(
      {
        eventType: 'ui',
        action: 'clicked',
        actionSubject: 'button',
        actionSubjectId: 'cancelButton',
        attributes: {
          annotated: hasBeenEdited,
          input,
        },
      },
      this.props.createAnalyticsEvent,
    );

    if (hasBeenEdited) {
      this.setState({ closeIntent: true });
    } else if (onClose) {
      onClose();
    }
  };

  onError = (error: any) => {
    this.setState({
      hasError: true,
      errorMessage: error,
    });
  };

  private clickShellNotPass = (e: React.SyntheticEvent<HTMLDivElement>) => {
    // Stop click from propagating back to the editor.
    // Without it editor will get focus and apply all the key events
    e.stopPropagation();
  };

  renderLoading = () => {
    return (
      <SpinnerWrapper>
        <Spinner size="large" invertColor={true} />
      </SpinnerWrapper>
    );
  };

  renderEditor = (imageUrl: string) => {
    const EditorViewWithAnalyticsContext = withAnalyticsContext({
      attributes: {
        annotated: this.state.hasBeenEdited,
      },
    })(EditorView);

    return (
      <EditorViewWithAnalyticsContext
        imageUrl={imageUrl}
        onSave={this.onSave}
        onCancel={this.onCancel}
        onError={this.onError}
        onAnyEdit={this.onAnyEdit}
      />
    );
  };

  renderError = (error: any) => {
    const { onClose } = this.props;
    if (error instanceof Error) {
      error = error.message;
    }
    return <ErrorView message={error} onCancel={onClose || (() => {})} />;
  };

  render() {
    const { imageUrl, hasError, errorMessage } = this.state;

    const content = hasError
      ? this.renderError(errorMessage)
      : imageUrl
      ? this.renderEditor(imageUrl)
      : this.renderLoading();

    return (
      <Blanket onClick={this.clickShellNotPass}>
        {this.renderDeleteConfirmation()}
        <Shortcut keyCode={27} handler={() => this.onCancel('esc')} />
        {content}
      </Blanket>
    );
  }
}

export default class extends React.Component<SmartMediaEditorProps> {
  render() {
    const Component = withAnalyticsContext({
      packageName,
      packageVersion,
    })(withAnalyticsEvents()(injectIntl(SmartMediaEditor)));
    const content = <Component {...this.props} />;
    return this.context.intl ? (
      content
    ) : (
      <IntlProvider locale="en">{content}</IntlProvider>
    );
  }
}
