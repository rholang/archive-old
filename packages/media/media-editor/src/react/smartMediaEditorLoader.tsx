import * as React from 'react';
import { ModalSpinner } from '@atlaskit/media-ui';
import { WithMediaClientConfigProps } from '@atlaskit/media-client';
import { N700A } from '@atlaskit/theme/colors';
import { SmartMediaEditorProps } from './smartMediaEditor';

type SmartEditorWithMediaClientConfigProps = WithMediaClientConfigProps<
  SmartMediaEditorProps
>;
type SmartEditorWithMediaClientConfigComponent = React.ComponentType<
  SmartEditorWithMediaClientConfigProps
>;

interface AsyncSmartMediaEditorState {
  SmartMediaEditor?: SmartEditorWithMediaClientConfigComponent;
}

export default class AsyncSmartMediaEditor extends React.PureComponent<
  SmartEditorWithMediaClientConfigProps & AsyncSmartMediaEditorState,
  AsyncSmartMediaEditorState & { isErrored: boolean }
> {
  static displayName = 'AsyncSmartMediaEditor';
  static SmartMediaEditor?: SmartEditorWithMediaClientConfigComponent;

  state = {
    // Set state value to equal to current static value of this class.
    SmartMediaEditor: AsyncSmartMediaEditor.SmartMediaEditor,
    isErrored: false,
  };

  async UNSAFE_componentWillMount() {
    if (!this.state.SmartMediaEditor) {
      try {
        const [mediaClient, smartEditorModule] = await Promise.all([
          import(/* webpackChunkName:"@atlaskit-media-client" */ '@atlaskit/media-client'),
          import(/* webpackChunkName:"@atlaskit-internal_smart-media-editor" */ './smartMediaEditor'),
        ]);
        AsyncSmartMediaEditor.SmartMediaEditor = mediaClient.withMediaClient(
          smartEditorModule.default,
        );
        this.setState({
          SmartMediaEditor: AsyncSmartMediaEditor.SmartMediaEditor,
        });
      } catch (e) {
        // eslint-disable-next-line no-console
        console.error(e);
        this.setState({ isErrored: true });
      }
    }
  }

  render() {
    const { isErrored } = this.state;

    if (isErrored) {
      return null;
    }
    if (!this.state.SmartMediaEditor) {
      return <ModalSpinner blankedColor={N700A} invertSpinnerColor={true} />;
    }

    return <this.state.SmartMediaEditor {...this.props} />;
  }
}
