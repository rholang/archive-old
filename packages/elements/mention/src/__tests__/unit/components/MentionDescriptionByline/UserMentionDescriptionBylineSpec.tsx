import { render } from 'enzyme';
import * as React from 'react';
import UserMentionDescriptionByline from '../../../../components/MentionDescriptionByline';
import { userMention } from './_commonData';
import { MentionDescription } from '../../../..';

const shallowRender = (mention: MentionDescription) =>
  render(<UserMentionDescriptionByline mention={mention} />);

describe('User mention description', () => {
  it('should render User Mention description component', () => {
    const component = shallowRender(userMention);
    expect(component).toMatchSnapshot();
  });

  it('should not show anything if name and nickname match', () => {
    const component = shallowRender({
      id: '12345',
      avatarUrl: 'www.example.com/image.png',
      name: 'Nickname',
      nickname: 'Nickname',
    });
    expect(component.text()).not.toContain('Nickname');
  });

  it('should show if name and nickname are different', () => {
    const component = shallowRender({
      id: '12345',
      avatarUrl: 'www.example.com/image.png',
      name: 'Different full name',
      nickname: 'Nickname',
    });
    expect(component.text()).toContain('Nickname');
  });

  it('should show if name and nickname have non-matching case', () => {
    const component = shallowRender({
      id: '12345',
      avatarUrl: 'www.example.com/image.png',
      name: 'Nickname',
      nickname: 'nickname',
    });
    expect(component.text()).toContain('nickname');
  });

  it('should show if name and nickname vary in whitespace', () => {
    const component = shallowRender({
      id: '12345',
      avatarUrl: 'www.example.com/image.png',
      name: 'Nick name',
      nickname: 'Nickname',
    });
    expect(component.text()).toContain('Nickname');
  });
});
