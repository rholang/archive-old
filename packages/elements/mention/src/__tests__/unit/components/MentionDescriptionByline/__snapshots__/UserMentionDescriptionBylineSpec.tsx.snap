// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User mention description should render User Mention description component 1`] = `
.c0 {
  color: #7A869A;
  font-size: 12px;
  margin-top: 2px;
  display: block;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

<span
  class="c0"
>
  @Test User
</span>
`;
